### ISSUE TITLE
`nimble install --link` as replacement for "old" (<=0.13.1) `nimble develop`

### ISSUE THREAD

**Issue by pietroppeter** [nim-lang/nimble#948]
with https://github.com/nim-lang/nimble/pull/913 (currently merged and slated for release in 0.14) `nimble develop` changed considerably its behaviour. as discussed in the forum (see [this](https://forum.nim-lang.org/t/8404#54212), [this](https://forum.nim-lang.org/t/8404#54460) and [this](https://forum.nim-lang.org/t/8404#54524) comments of the same thread discussing Nim 1.6.0 RC1) a viable option to keep supporting the workflows that old `nimble develop` allowed is to replicate its functionality in a `nimble install --link` command.

as an example workflow that was supported by old `nimble develop` (and would keep being supported by this `nimble install --link` is the following):
* you have nim package `mylib`, you clone it and put it in develop mode
* in another folder you have a standalone nim file (`standalone.nim`) that reproduces a bug happening in `mylib`. this standalone file has no project around it nor a nimble file.
* you run `nim r standalone` and the bug reproduces.
* you modify `mylib` and try again `nim r standalone`, maybe your bug is fixed, maybe not and you can go back and forth in modifying `mylib` changing stuff and adding echo statements and always running only `nim r standalone` until you are satisfied with this change

how did old `nimble.develop` was able to support this workflow?

- it [created a special `nimble-link` file](https://github.com/nim-lang/nimble/blob/d13f3b8ce288b4dc8c34c219a4e050aaeaf43fc9/src/nimble.nim#L985) inside folder `~/.nimble/pkgs/mylib#head` that looked like this:
```
path-to-my-repo\mylib\mylib.nimble
path-to-my-repo\mylib\src
```
- [nim is aware of this specific `nimble-link` format](https://github.com/nim-lang/Nim/blob/5e6680406f39b2b0b90d8b137285e066cc358d43/compiler/nimblecmd.nim#L126) and is able to follow the link

**Comment by arnetheduck** [comment]
One important point, and one source of endless debugging and user frustration that we've experienced, is when users forget that there's a global develop package and it messes up their builds, causing them to raise issues with unrelated projects. Nearly every developer that we on-board to Nim runs into this problem.

This is specially true when working on anything beyond a single project.

This is one among many reasons why we've been pushing to deprecate global develop modes as far as possible - it's a bad design that leads to bad outcomes for the whole community - the UX is simply sub-par in the long run and while it scores some points on my-first-nim-program, that score and much more is quickly lost on my-second-nim-program. 

If magic globals are to remain within close user reach in Nimble (even for a transition period), they would ideally be clearly flagged, explicit, and every time there's a build involving them, the build output should contain information about these globals potentially breaking it.

One way to do this to rename any `nimble develop` emulation, then turn `nimble develop` into a help text that explains how to get back the old behavior while adapting to the new work flow.

**Comment by bobeff** [comment]
@pietroppeter @Araq @dom96 @arnetheduck

One possible solution is to have a global develop file for which Nim to be aware. When you execute `nimble develop --global <some_pkg>` the path to the package will be written in the global develop file and it will be used only when there is no local develop file. One drawback of this solution is that we have to make Nim aware of this global develop file and to be able to parse it and use written in it paths to packages.

Another solution is to have a third directory alongside `pkgs` and `pkgs2`. Let say that it is called `links`  and there to be written packages links to the packages directories which to be used only when there is no local develop file. A drawback of this solution is that it will not work with the lock files and sync functionality.

**Comment by dom96** [comment]
Why don't we simply include the old implementation under a different command (or ideally under the same command) and call it a day?

**Comment by bobeff** [comment]
@dom96 Where to create the link file?

- If I create it in the `pkgs2` directory it will be not correct because the new format of the packages directories requires the contents to correspond to the checksum in their names.

- If I create it in the `pkgs` directory it will be not correct because of the new format of the package metadata. When the user switches to old Nimble, the old Nimble will not be able to read the metadata from the directory.

For those reasons, I propose a third directory as a possible solution.

**Comment by dom96** [comment]
Instead of creating yet another directory, how about we just put these into the `pkgs2` directory and have the checksum calculation ignore those files?

The real question is how should these old-fashioned links interact with how Nimble resolves deps? should they override all the locked deps? should we force users to use the new `develop` mode when working with lock files? These are far more important questions that we should discuss.

**Comment by bobeff** [comment]
I think that those links will have to work only when directly invoking the compiler with a standalone `.nim` file. Nimble can simply ignore them. If you want to use a develop mode package with another Nimble package you should use the new develop mode.

**Comment by beef331** [comment]
Why not instead of disallowing any global packages with nimble commands, just make it so `nimble` can take a `--globalOverrides` or similar which defers to global linked when one is linked instead of locked/versioned one. So `nimble develop --link` or w/e is decided on does what `nimble develop` does presently, and then you do `nimble --globalOverrides run` to run the program with the develop packages. This addresses arne's concern and also maintains the more inuitive dev process when you have many packages you want to work in which are related.

**Comment by bobeff** [comment]
I did the change that the links files are treated as a global develop file used only when no local one is found. This way they should work seamlessly with the lock files and other develop files.

**Comment by pietroppeter** [comment]
> One possible solution is to have a global develop file for which Nim to be aware. When you execute nimble develop --global <some_pkg> the path to the package will be written in the global develop file and it will be used only when there is no local develop file. One drawback of this solution is that we have to make Nim aware of this global develop file and to be able to parse it and use written in it paths to packages.

> Another solution is to have a third directory alongside pkgs and pkgs2. Let say that it is called links and there to be written packages links to the packages directories which to be used only when there is no local develop file. A drawback of this solution is that it will not work with the lock files and sync functionality. Nim also has to be aware of the link format.

I would personally be happy with any of those two solutions, if I understand correctly both provide same behaviour as before when compiling/running with `nim c/r` after a package has been made in `nimble develop --global`. I see no problem in ignoring the global develop file when a local one is present. I also do not see particular issues in ignoring it also for builds with nimble.

When I opened this issue I thought it would just be a simple matter of adding the old feature back, but as you [clearly explain](https://github.com/nim-lang/nimble/issues/948#issuecomment-997915537) this would not have worked.

I understand from commits and further discussion that you already picked the global develop file option, correct?

Is it correct that if this is approved and merged you would also need to made a related PR in Nim compiler for this to work? Would this then imply that in order for new nimble global develop to work you would need to have a recent Nim (I guess the fix could be backported to previous Nim versions)?

**Comment by bobeff** [comment]
@pietroppeter 

> I understand from commits and further discussion that you already picked the global develop file option, correct?

No. I picked the solution with additional directory `links` and the old link files in it. I managed to make the solution work with the link files as links in a develop file so it has to work transparently with the lock files.

> Is it correct that if this is approved and merged you would also need to made a related PR in Nim compiler for this to work?

Yes. It is a simple addition to the `nim.cfg` file which adds the new `links` directory to the Nimble path.

> Would this then imply that in order for new nimble global develop to work you would need to have a recent Nim (I guess the fix could be backported to previous Nim versions)?

Yes, it requires a recent Nim and I see no obstacle to backport it just like for the `pkgs2` directory.
